internal/provider/segment_resource.go

```diff
@@ -0,0 +1,159 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-provider-grafana-adaptive-metrics/internal/client"
+	"github.com/hashicorp/terraform-provider-grafana-adaptive-metrics/internal/model"
+)
+
+type segmentResource struct {
+	client *client.Client
+}
+
+var (
+	_ resource.Resource                = &segmentResource{}
+	_ resource.ResourceWithConfigure   = &segmentResource{}
+	_ resource.ResourceWithImportState = &segmentResource{}
+)
+
+func newSegmentResource() resource.Resource {
+	return &segmentResource{}
+}
+
+func (e *segmentResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	data, ok := req.ProviderData.(*resourceData)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected resource configure type",
+			fmt.Sprintf("Got %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+
+	e.client = data.client
+}
+
+func (e *segmentResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = fmt.Sprintf("%s_segment", req.ProviderTypeName)
+}
+
+func (e *segmentResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Description: "A UILD that uniquely identifies the segment.",
+			},
+			"name": schema.StringAttribute{
+				Required:    true,
+				Description: "The name of the segment.",
+			},
+			"selector": schema.StringAttribute{
+				Required:    true,
+				Description: "The selector that defines the segment.",
+			},
+			"fallback_to_default": schema.BoolAttribute{
+				Optional:    true,
+				Computed:    true,
+				Default:     defaultBoolTrue{},
+				Description: "Whether to fallback to the default segment if the selector does not match any segments.",
+			},
+		},
+	}
+}
+
+func (e *segmentResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
```

Comment:
The four methods `Create()` `Read()` `Update()` `Delete()` deal with a `model.Segment` and a `model.SegmentTF`. Could we keep their variable names consistent? Currently they're named:

`s` / `state`
`ex` / `tf`
`ex` / `state`
N/A / `state`

Reply:
Yep, renamed so that `segment` is always a `model.Segment`. And `plan` or `state` are both `model.SegmentTF`, since plan and state are terraform concepts.